---
title: "Partitioned Local Depth (PaLD) Clustering Analyses in R"
author:
  - name: Lucy D'Agostino McGowan
    affiliation: Wake Forest University
    address:
    - Winston-Salem, NC
    - 27106
    email:  mcgowald@wfu.edu
  - name: Katherine Moore
    affiliation: Wake Forest Unversity
    address:
    - Winston-Salem, NC
    - 27106
    email: mooreke@wfu.edu 
  - name: Kenneth Berenhaut 
    affiliation: Wake Forest University
    address:
    - Winston-Salem, NC
    - 27106
    email: berenhks@wfu.edu 
abstract: >
  An abstract of less than 150 words.
output:
  rticles::rjournal_article:
    includes:
      in_header: preamble.tex
---


```{r, echo = FALSE}
knitr::opts_chunk$set(fig.height = 4, message = FALSE, warning = FALSE)
```

## Introduction

* Describe PaLD (Ken & Kate?) [@berenhaut2022social]

We present a new package, \CRANpkg{pald}, for calculating partitioned local depth (PaLD) probabilities, implementing clustering analyses, and creating data visualizations to represent the clusters. This paper will describe how to use the package as well as walk through two examples.

## pald

The main functions in \CRANpkg{pald} package can be split into 3 categories:

1. Helper functions to organize data into the correct format, into distance matrices and then cohesion matrices
2. Functions that convert a cohesion matrix into a variety of useful formats, including partitioned local depths, clusters, and graphs
3. Plotting functions

In addition, the package provides a number of pertinent example data sets commonly used to demonstrate cluster algorithms, including a synthetic data set of two-dimensional points created by \citet{gionis1clustering} to demonstrate clustering aggregation, a sample of walking distances from a pump in the infamous cholera outbreak \citep{cholera}, clustering data generated from the scikit-learn Python package \citep{pedregosa2011scikit}, and data compiled by \cite{tissue} of tissue gene expressions. 

While it is not a necessity, the \CRANpkg{pald} package is designed to function well with the pipe operator, `|>`. This functionality will be demonstrated below.


### Helper functions to create contribution matrix

For demonstration purposes, below is a sample data frame with two variables, `x1` and `x2`. The methods put forth here work on data frames with higher dimensions, as described in the **Examples** section; we are simply choosing a small data frame here for demonstration purposes.

```{r}
library(pald)
df <- data.frame(
  x1 = c(6, 8, 11, 16, 4),
  x2 = c(5, 4, 13, 7, 18)
)
rownames(df) <- c("A", "B", "C", "D", "E")
```

The first step needed to calculate the partitioned local depths is to construct a _distance matrix_. If the data are already in this form, the user can skip to the next step. The `dist()` function converts an input data frame into a distance matrix, as demonstrated below.

```{r}
d <- dist(df)
```

This will create an $n\times n$ distance matrix, where $n$ corresponds to the number of observations in the original data frame, in this example $n = 5$. 
<!-- TODO: is this nxn? looks like (n-1)x(n-1) -->
This distance matrix can then be passed to the `cohesion_matrix()` function in order to calculate the pairwise cohesion values. Cohesion is an interpretable probability that reflects the strength of alignment of two points. Again, if the user begins with a distance matrix, they can skip the first step and simply input the distance matrix into this function.

```{r}
d <- dist(df)
cohesion_matrix(d)
```

Equivalently, the user can use the native pipe `|>` as follows.

```{r}
df |>
  dist() |>
  cohesion_matrix()
```

The _cohesion matrix_ output by the `cohesion_matrix()` is the main input for the majority of the remaining functions.

### Functions that convert a cohesion matrix into useful formats

From the _cohesion matrix_, a variety of useful quantities can be calculated. Below, we create a cohesion matrix using the functions described in the previous section.

```{r}
d |>
  dist() |>
  cohesion_matrix() -> cohesion
```

To calculate the _clusters_ that each point will fall into, we can use the `community_clusters()` function. This will output a data frame with two columns, the first will correspond to the `point`, as identified by the row name of the original input data frame, `df`, the second will identify the `cluster` that each point belongs to.

```{r}
community_clusters(cohesion)
```

In this example, three clusters are identified with these five points. Points `A` and `B` fall into cluster 1. Points `C` and `D` into cluster 2, and point `E` in cluster 3.

The `local_depths()` function calculates the _depths_ of each point, outputting a vector of local depths. Local depth is an interpretable probability which reflects aspects of relative position and centrality via distance comparisons (i.e., $d(z, x) < d(z, y)$).

```{r}
local_depths(cohesion)
```

In this case, the deepest point is `C`. 

The `strong_threshold()` function will calculate the cohesion threshold for strong ties. This is equal to half the average of the diagonal cohesion matrix. This is a threshold that may be used to distinguish between strong and weak ties.

```{r}
strong_threshold(cohesion)
```

In this case, the threshold is `0.14`.

The `any_isolated()` function will check whether there are any isolated points that will inadvertently be dropped by a graph.

```{r}
any_isolated(cohesion)
```

Here, there are no isolated points.

The function `cohesion_strong()` will update the cohesion matrix to set all weak ties to zero (via the `strong_threshold()` function). Optionally, the matrix will also be symmetrized, with the default parameter `symmetric = TRUE`.

```{r}
cohesion_strong(cohesion)
```

Finally, the `community_graphs()` function takes the cohesion matrix and creates \CRANpkg{igraph} objects, graphs that describes the relationship between the points. This function will output a list of three objects:

* `G`: the weighted (community) graph whose edge weights are mutual cohesion
* `G_strong`: the weighted (community) graph consisting of edges for which mutual cohesion is greater than the threshold for strong ties
* `layout`: the graph layout, using the Fruchterman Reingold (FR) force-directed graph drawing for the graph `G`

```{r}
graphs <- community_graphs(cohesion)
graphs[["G_strong"]]
```

Here we see that there are two connected components, points `A` and `B`, which form the first cluster, and points `D` and `C` which form the second.

## Plotting functions

The final category of function is functions for data visualization. We can begin by visualizing the points in data frame `df` (Figure \ref{fig:fig1}). 

```{r fig1, fig.cap = "Visualize the points from data frame `df`", message = FALSE, warning = FALSE}
library(ggplot2)
ggplot(df, aes(x1, x2)) +
  geom_text(label = rownames(df))
```

We can then pass the cohesion matrix to the `plot_community_graphs()` function to view the relationship between points (Figure \ref{fig:fig2}).

```{r fig2, fig.cap = "PaLD graph displaying the relationship between the points in data frame `df`"}
df |>
  dist() |>
  cohesion_matrix() |>
  plot_community_graphs()
```

The `layout` argument allows the user to pass a matrix to dictate the layout of the graph. For example, if we wanted the graph to match the visualization displayed in Figure \ref{fig:fig1}, we can pass `as.matrix(df)`, or a matrix of the data frame `df` to the `layout` argument  (Figure \ref{fig:fig3}.  
```{r fig3, fig.cap = "PaLD graph displaying the relationship between the points in data frame `d`, matching the original layout in Figure 1"}
df |>
  dist() |>
  cohesion_matrix() |>
  plot_community_graphs(layout = as.matrix(df))
```

This `plot_community_graphs()` function will also permit parameters that can be passed to the `plot.igraph()` function. For example, to add axes to the graph, the user can pass the `axes = TRUE` argument to the `...` in the `plot_community_graphs()` function (Figure \ref{fig:fig4}).

```{r fig4, fig.cap = "PaLD graph displaying the relationship between the points in data frame `d`, matching the original layout in Figure 1, adding axes"}
df |>
  dist() |>
  cohesion_matrix() |>
  plot_community_graphs(layout = as.matrix(df),
                        axes = TRUE)
```

## Examples

We will demonstrate the utility of the \CRANpkg{pald} package in two clustering examples.

### Clustering tissue gene expression data

The first example if from a subset of tissue gene expression data from \citet{zilliox2007gene}, \citet{mccall2011gene}, and \citet{mccall2014gene}, obtained from the **tissuesGeneExpression** bioconductor package \citep{tissue}. A cohesion matrix was created using this data set and is included the \CRANpkg{pald} package in an object called `tissue_cohesion_matrix`.

The `tissue_cohesion_matrix` object is a cohesion matrix with 189 rows and 189 columns.

We can use this contribution matrix to display the relationship between tissue samples using the `plot_community_graphs()` function (Figure \ref{fig:fig5}). For clarity of the display, we show how to only include a random set of the vertex labels. We can pass this random set through the `...` to the `plot.igraph` `vertex.label` parameter. 

```{r fig5, fig.cap = "PaLD clustering of tissue data", fig.height = 9, fig.width = 7}
set.seed(1)

labels <- rownames(tissue_cohesion_matrix)
labels[sample(1:189, 125)] <- ""
plot_community_graphs(tissue_cohesion_matrix,
                      vertex.label = labels,
                      vertex.size = 4)
```

The `community_clusters()` function can be used to identify the clusters of each tissue sample. Since the output is a data frame, we can summarize the clusters using commonly used data analysis techniques. For demonstration purposes, we will use the \CRANpkg{dplyr} package to summarize the contribution of clusters.

```{r}
library(dplyr)
community_clusters(tissue_cohesion_matrix) %>%
  group_by(cluster, point) %>%
  count()
```

From this, we can glean that cluster one consists of two types of tissue, the kidney and endometrium. Cluster two is comprised of only the hippocampus. 

### Clustering generated data

The \CRANpkg{pald} includes three randomly generated data frames corresponding to plots from @berenhaut2022social: 

* `exdata1` is a data set consisting of 8 points to recreate Figure 1 in @berenhaut2022social 
* `exdata2` is a data set consisting of 16 points to recreate Figure 2 in @berenhaut2022social 
* `exdata3` is a data set consisting of 240 points to recreate Figure 4D in @berenhaut2022social 

Here, we will demonstrate how to use `exdata3`. These points were generated from bivariate normal distributions with varying means and variances.

```{r}
exdata_cohesion_matrix <- exdata3 |>
  dist() |>
  cohesion_matrix()
```

When plotting the `exdata3` PaLD graph, we want the layout to match the layout of the original data, so we will pass `as.matrix(exdata3)` to the `layout` parameter in the `plot_community_graphs()` function. Additionally, here the row names are meaningless, they just correspond to the location of the generated data, so we can remove the labels on the plot by setting `show_labels = FALSE`. The `only_strong` parameter will only display the strongly connected edges. Additionally, we can optionally pass parameters through the `...` to `plot::igraph` function; for example, to increase the vertex size we can set `vertex.size = 5`. Figure \ref{fig:fig6} shows this.

```{r fig6, fig.cap = "PaLD clustering of randomly generated example data (from Figure 4D from Berenhaut et al. (2022))"}
plot_community_graphs(exdata_cohesion_matrix,
                layout = as.matrix(exdata3),
                show_labels = FALSE, 
                only_strong = TRUE,
                vertex.size = 5)
```

The ability of the PaLD algorithm to discern clusters is demonstrated here.

## Summary

This paper introduces the \CRANpkg{pald} package, demonstrating it's utility for providing a parameter-free clustering algorithm that can easily be applied to any data set.

\bibliography{RJreferences}
